includes/class-shortcodes.php

```diff
@@ -0,0 +1,486 @@
+			}
+
+			// Add in the more indicator.
+			$excerpt = $excerpt . $excerpt_more;
+
+			// Run it through the excerpt filter which will add some html tags back in.
+			$excerpt_filtered = apply_filters( 'the_excerpt', $excerpt );
+
+			// Now set the current excerpt length to this new filtered length.
+			$current_excerpt_length = strlen( $excerpt_filtered );
+
+			// Check to see if we're over the target length.
+			if ( $current_excerpt_length > $target_excerpt_length ) {
+				// If so, remove 20 characters from the current max and run the loop again.
+				$current_excerpt_max = $current_excerpt_max - 20;
+			}
+		}
+
+		return \apply_filters( 'the_excerpt', $excerpt );
+	}
+
+	/**
+	 * Generates output for the ap_content shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function content( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		$content = \get_post_field( 'post_content', $this->post );
+
+		return \apply_filters( 'the_content', $content );
+	}
+
+	/**
+	 * Generates output for the ap_permalink shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function permalink( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		return \sprintf( '<a href="%1$s">%1$s</a>', \esc_url( \get_permalink( $this->post->ID ) ) );
+	}
+
+	/**
+	 * Generates output for the ap_shortlink shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function shortlink( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		return \sprintf( '<a href="%1$s">%1$s</a>', \esc_url( \wp_get_shortlink( $this->post->ID ) ) );
+	}
+
+	/**
+	 * Generates output for the ap_thumbnail shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function thumbnail( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		$image = \get_the_post_thumbnail_url( $this->post->ID, 'thumbnail' );
+
+		if ( ! $image ) {
+			return '';
+		}
+
+		return $image;
+	}
+
+	/**
+	 * Generates output for the ap_image shortcode
+	 *
+	 * @param array  $atts 		shortcode attributes
+	 * @param string $content 	shortcode content
+	 * @param string $tag 		shortcode tag name
+	 *
+	 * @return string
+	 */
+	public function image( $atts, $content, $tag ) {
+		if( $this->post === false ) {
+			return '';
+		}
+
+		$size = 'full';
+
+		if( is_array( $atts ) && array_key_exists( 'size', $atts ) ) {
```

Comment:
use `shortcode_atts` to set default params: https://developer.wordpress.org/reference/functions/shortcode_atts/

Reply:
Done.